src/nomad_simulations/schema_packages/model_system.py

```diff
@@ -1127,9 +1150,78 @@ def resolve_system_type_and_dimensionality(
 
         return system_type, dimensionality
 
+    def _walk_depth_first(self):
+        """Yield self and every descendant ModelSystem (depth-first)."""
+        yield self
+        for sub in self.sub_systems:
+            yield from sub._walk_depth_first()
+
+    def _consolidate_particle_states(self, logger):
+        """
+        Make sure the representative (top-level) ModelSystem owns a complete,
+        ordered ``particle_states`` list, even if only descendants populated it.
+        """
+        if not self.is_representative:
+            return  # run only on the root
+
+        # 1. Determine how many atoms we expect
+        n_atoms = self.n_particles
+        if n_atoms is None:
+            max_idx = -1
+            for ms in self._walk_depth_first():
+                if ms.particle_indices:
+                    max_idx = max(max_idx, *ms.particle_indices)
+            n_atoms = max_idx + 1
+
+        if n_atoms <= 0:
+            logger.debug('No particles detected; nothing to consolidate.')
+            return
+
+        # 2. Ensure self.particle_states has the right length
+        if not self.particle_states:
+            self.particle_states = [None] * n_atoms
+        elif len(self.particle_states) < n_atoms:
+            self.particle_states.extend([None] * (n_atoms - len(self.particle_states)))
```

Comment:
I think these blocks can be removed. Their logic serves 2 minor purposes:

- ensuring the population of `n_atoms`. This can be done _independently_ during normalization.
- presetting the root's particle_states, which can be done by simply returning the constructed `list` and assigning it (if non-empty) in the normalization body.